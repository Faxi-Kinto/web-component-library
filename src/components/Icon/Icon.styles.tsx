import styled from 'styled-components';
import { flexRow } from '../../styles/flex';
import { pxToRem } from '../../styles/basics';
import { defaultFontSize } from '../../styles/constants';

type ContainerProps = {
  size?: string;
};

export const Container = styled.div`
  display: flex;
  ${(props: ContainerProps): string =>
    props.size
      ? `font-size: ${props.size}`
      : `font-size: ${pxToRem(defaultFontSize)}`};
`;

export const IconSet = styled.div`
  ${flexRow()};

  > * {
    margin: ${pxToRem('10px')};
  }
`;
