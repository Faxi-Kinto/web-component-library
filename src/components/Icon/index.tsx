import {
  library,
  IconProp,
  IconDefinition,
  IconPack,
} from '@fortawesome/fontawesome-svg-core';

export const setLibraries = (
  libraries: (IconDefinition | IconPack)[]
): void => {
  library.add(...libraries);
};

export let mapNamePropToFaNames = <T,>(_iconName: T): IconProp => {
  throw Error('Please provide your function');
};

export const setMapNamePropToFaNames = <T,>(
  customMapNamePropToFaNames: (iconName: T) => IconProp
): void => {
  mapNamePropToFaNames = customMapNamePropToFaNames as <T>(
    iconName: T
  ) => IconProp;
};
