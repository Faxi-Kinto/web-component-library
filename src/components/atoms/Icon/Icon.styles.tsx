import styled from 'styled-components';
import { flexRow } from '@faxi/web-css-utilities';
import { pxToRem, size } from '@faxi/web-css-utilities';
import { defaultFontSize } from '@faxi/web-css-utilities/constants';

type ContainerProps = {
  size?: string;
};

export const Container = styled.div`
  ${flexRow('center', 'center')};
  ${(props: ContainerProps): string =>
    props.size
      ? `${size(props.size)}; font-size: ${props.size}`
      : `${size(pxToRem(defaultFontSize))}; font-size: ${pxToRem(
          defaultFontSize
        )}`};
`;

export const IconSet = styled.div`
  ${flexRow()};

  > * {
    margin: ${pxToRem('10px')};
  }
`;
