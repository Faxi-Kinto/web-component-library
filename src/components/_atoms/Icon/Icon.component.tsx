import * as React from 'react';
import * as Styled from './Icon.styles';
import { FontAwesomeIcon } from '@fortawesome/react-fontawesome';
import { mapNamePropToFaNames } from '../../Icon';
/**
 * @name Icon
 *
 * @returns {JSX}
 */

export type IconProps<T> = {
  name: T;
  size?: string;
  color?: string;
  className?: string;
  onClick?: (event: React.MouseEvent<HTMLDivElement>) => void;
  id?: string;
};

const Icon = <T,>(props: IconProps<T>): JSX.Element => {
  const { size, color, name, className, onClick, id } = props;

  const iconName = mapNamePropToFaNames<T>(name);

  return (
    <Styled.Container
      id={id}
      size={size}
      className={className}
      onClick={onClick}
    >
      <FontAwesomeIcon color={color} size={'1x'} icon={iconName} />
    </Styled.Container>
  );
};

export default Icon;
