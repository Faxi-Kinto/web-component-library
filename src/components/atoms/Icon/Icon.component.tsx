import * as React from 'react';
import * as Styled from './Icon.styles';
import { FontAwesomeIcon } from '@fortawesome/react-fontawesome';
import { mapNamePropToFaNames } from '../../../setupFontAwesome';
import { IconProp } from '@fortawesome/fontawesome-svg-core';

/**
 * @name Icon
 *
 * [Insert Component Description]
 *
 * @returns {JSX}
 */

export type IconProps = {
  name: IconProp;
  size?: string;
  color?: string;
  className?: string;
};

const Icon: React.FC<IconProps> = (props: IconProps): JSX.Element => {
  const { size, color, name, className } = props;

  const iconName = mapNamePropToFaNames(name);

  return (
    <Styled.Container size={size} className={className}>
      <FontAwesomeIcon color={color} size={'1x'} icon={iconName} />
    </Styled.Container>
  );
};

export default Icon;
