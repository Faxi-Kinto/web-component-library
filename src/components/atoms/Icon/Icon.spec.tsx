import React from 'react';
import renderer from 'react-test-renderer';
import Icon from './Icon.component';

test('Icon renders correctly', () => {
  const component = renderer.create(<Icon name="users" />);

  const tree = component.toJSON();

  expect(tree).toMatchSnapshot();
});
