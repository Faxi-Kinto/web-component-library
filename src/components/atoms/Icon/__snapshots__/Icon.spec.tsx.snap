// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`All Icons render correctly 1`] = `
.c1 {
  display: -webkit-box;
  display: -webkit-flex;
  display: -ms-flexbox;
  display: flex;
  -webkit-flex-direction: row;
  -ms-flex-direction: row;
  flex-direction: row;
  -webkit-box-pack: center;
  -webkit-justify-content: center;
  -ms-flex-pack: center;
  justify-content: center;
  -webkit-align-items: center;
  -webkit-box-align: center;
  -ms-flex-align: center;
  align-items: center;
  width: 1rem;
  height: 1rem;
  font-size: 1rem;
}

.c0 {
  display: -webkit-box;
  display: -webkit-flex;
  display: -ms-flexbox;
  display: flex;
  -webkit-flex-direction: row;
  -ms-flex-direction: row;
  flex-direction: row;
  -webkit-box-pack: initial;
  -webkit-justify-content: initial;
  -ms-flex-pack: initial;
  justify-content: initial;
  -webkit-align-items: initial;
  -webkit-box-align: initial;
  -ms-flex-align: initial;
  align-items: initial;
}

.c0 > * {
  margin: 0.625rem;
}

<div
  className="c0"
>
  <div
    className="c1"
    size="1rem"
  >
    <svg
      aria-hidden="true"
      className="svg-inline--fa fa-plus fa-w-12 fa-1x "
      data-icon="plus"
      data-prefix="fal"
      focusable="false"
      role="img"
      style={Object {}}
      viewBox="0 0 384 512"
      xmlns="http://www.w3.org/2000/svg"
    >
      <path
        d="M376 232H216V72c0-4.42-3.58-8-8-8h-32c-4.42 0-8 3.58-8 8v160H8c-4.42 0-8 3.58-8 8v32c0 4.42 3.58 8 8 8h160v160c0 4.42 3.58 8 8 8h32c4.42 0 8-3.58 8-8V280h160c4.42 0 8-3.58 8-8v-32c0-4.42-3.58-8-8-8z"
        fill="currentColor"
        style={Object {}}
      />
    </svg>
  </div>
  <div
    className="c1"
    size="1rem"
  >
    <svg
      aria-hidden="true"
      className="svg-inline--fa fa-users fa-w-20 fa-1x "
      data-icon="users"
      data-prefix="fal"
      focusable="false"
      role="img"
      style={Object {}}
      viewBox="0 0 640 512"
      xmlns="http://www.w3.org/2000/svg"
    >
      <path
        d="M544 224c44.2 0 80-35.8 80-80s-35.8-80-80-80-80 35.8-80 80 35.8 80 80 80zm0-128c26.5 0 48 21.5 48 48s-21.5 48-48 48-48-21.5-48-48 21.5-48 48-48zM320 256c61.9 0 112-50.1 112-112S381.9 32 320 32 208 82.1 208 144s50.1 112 112 112zm0-192c44.1 0 80 35.9 80 80s-35.9 80-80 80-80-35.9-80-80 35.9-80 80-80zm244 192h-40c-15.2 0-29.3 4.8-41.1 12.9 9.4 6.4 17.9 13.9 25.4 22.4 4.9-2.1 10.2-3.3 15.7-3.3h40c24.2 0 44 21.5 44 48 0 8.8 7.2 16 16 16s16-7.2 16-16c0-44.1-34.1-80-76-80zM96 224c44.2 0 80-35.8 80-80s-35.8-80-80-80-80 35.8-80 80 35.8 80 80 80zm0-128c26.5 0 48 21.5 48 48s-21.5 48-48 48-48-21.5-48-48 21.5-48 48-48zm304.1 180c-33.4 0-41.7 12-80.1 12-38.4 0-46.7-12-80.1-12-36.3 0-71.6 16.2-92.3 46.9-12.4 18.4-19.6 40.5-19.6 64.3V432c0 26.5 21.5 48 48 48h288c26.5 0 48-21.5 48-48v-44.8c0-23.8-7.2-45.9-19.6-64.3-20.7-30.7-56-46.9-92.3-46.9zM480 432c0 8.8-7.2 16-16 16H176c-8.8 0-16-7.2-16-16v-44.8c0-16.6 4.9-32.7 14.1-46.4 13.8-20.5 38.4-32.8 65.7-32.8 27.4 0 37.2 12 80.2 12s52.8-12 80.1-12c27.3 0 51.9 12.3 65.7 32.8 9.2 13.7 14.1 29.8 14.1 46.4V432zM157.1 268.9c-11.9-8.1-26-12.9-41.1-12.9H76c-41.9 0-76 35.9-76 80 0 8.8 7.2 16 16 16s16-7.2 16-16c0-26.5 19.8-48 44-48h40c5.5 0 10.8 1.2 15.7 3.3 7.5-8.5 16.1-16 25.4-22.4z"
        fill="currentColor"
        style={Object {}}
      />
    </svg>
  </div>
  <div
    className="c1"
    size="1rem"
  >
    <svg
      aria-hidden="true"
      className="svg-inline--fa fa-sliders-h fa-w-16 fa-1x "
      data-icon="sliders-h"
      data-prefix="far"
      focusable="false"
      role="img"
      style={Object {}}
      viewBox="0 0 512 512"
      xmlns="http://www.w3.org/2000/svg"
    >
      <path
        d="M496 72H288V48c0-8.8-7.2-16-16-16h-32c-8.8 0-16 7.2-16 16v24H16C7.2 72 0 79.2 0 88v16c0 8.8 7.2 16 16 16h208v24c0 8.8 7.2 16 16 16h32c8.8 0 16-7.2 16-16v-24h208c8.8 0 16-7.2 16-16V88c0-8.8-7.2-16-16-16zm0 320H160v-24c0-8.8-7.2-16-16-16h-32c-8.8 0-16 7.2-16 16v24H16c-8.8 0-16 7.2-16 16v16c0 8.8 7.2 16 16 16h80v24c0 8.8 7.2 16 16 16h32c8.8 0 16-7.2 16-16v-24h336c8.8 0 16-7.2 16-16v-16c0-8.8-7.2-16-16-16zm0-160h-80v-24c0-8.8-7.2-16-16-16h-32c-8.8 0-16 7.2-16 16v24H16c-8.8 0-16 7.2-16 16v16c0 8.8 7.2 16 16 16h336v24c0 8.8 7.2 16 16 16h32c8.8 0 16-7.2 16-16v-24h80c8.8 0 16-7.2 16-16v-16c0-8.8-7.2-16-16-16z"
        fill="currentColor"
        style={Object {}}
      />
    </svg>
  </div>
  <div
    className="c1"
    size="1rem"
  >
    <svg
      aria-hidden="true"
      className="svg-inline--fa fa-ballot fa-w-14 fa-1x "
      data-icon="ballot"
      data-prefix="fal"
      focusable="false"
      role="img"
      style={Object {}}
      viewBox="0 0 448 512"
      xmlns="http://www.w3.org/2000/svg"
    >
      <path
        d="M144 80h-32c-17.7 0-32 14.4-32 32v32c0 17.6 14.3 32 32 32h32c17.7 0 32-14.4 32-32v-32c0-17.6-14.3-32-32-32zm0 64h-32v-32h32v32zM416 0H32C14.3 0 0 14.4 0 32v448c0 17.6 14.3 32 32 32h384c17.7 0 32-14.4 32-32V32c0-17.6-14.3-32-32-32zm0 480H32V32h384v448zm-72-240H216c-4.4 0-8 3.6-8 8v16c0 4.4 3.6 8 8 8h128c4.4 0 8-3.6 8-8v-16c0-4.4-3.6-8-8-8zm-200 96h-32c-17.7 0-32 14.4-32 32v32c0 17.6 14.3 32 32 32h32c17.7 0 32-14.4 32-32v-32c0-17.6-14.3-32-32-32zm0 64h-32v-32h32v32zm200-32H216c-4.4 0-8 3.6-8 8v16c0 4.4 3.6 8 8 8h128c4.4 0 8-3.6 8-8v-16c0-4.4-3.6-8-8-8zm0-256H216c-4.4 0-8 3.6-8 8v16c0 4.4 3.6 8 8 8h128c4.4 0 8-3.6 8-8v-16c0-4.4-3.6-8-8-8zm-200 96h-32c-17.7 0-32 14.4-32 32v32c0 17.6 14.3 32 32 32h32c17.7 0 32-14.4 32-32v-32c0-17.6-14.3-32-32-32zm0 64h-32v-32h32v32z"
        fill="currentColor"
        style={Object {}}
      />
    </svg>
  </div>
  <div
    className="c1"
    size="1rem"
  >
    <svg
      aria-hidden="true"
      className="svg-inline--fa fa-map-marker-alt fa-w-12 fa-1x "
      data-icon="map-marker-alt"
      data-prefix="fal"
      focusable="false"
      role="img"
      style={Object {}}
      viewBox="0 0 384 512"
      xmlns="http://www.w3.org/2000/svg"
    >
      <path
        d="M192 96c-52.935 0-96 43.065-96 96s43.065 96 96 96 96-43.065 96-96-43.065-96-96-96zm0 160c-35.29 0-64-28.71-64-64s28.71-64 64-64 64 28.71 64 64-28.71 64-64 64zm0-256C85.961 0 0 85.961 0 192c0 77.413 26.97 99.031 172.268 309.67 9.534 13.772 29.929 13.774 39.465 0C357.03 291.031 384 269.413 384 192 384 85.961 298.039 0 192 0zm0 473.931C52.705 272.488 32 256.494 32 192c0-42.738 16.643-82.917 46.863-113.137S149.262 32 192 32s82.917 16.643 113.137 46.863S352 149.262 352 192c0 64.49-20.692 80.47-160 281.931z"
        fill="currentColor"
        style={Object {}}
      />
    </svg>
  </div>
  <div
    className="c1"
    size="1rem"
  >
    <svg
      aria-hidden="true"
      className="svg-inline--fa fa-comment-alt-lines fa-w-16 fa-1x "
      data-icon="comment-alt-lines"
      data-prefix="fal"
      focusable="false"
      role="img"
      style={Object {}}
      viewBox="0 0 512 512"
      xmlns="http://www.w3.org/2000/svg"
    >
      <path
        d="M448 0H64C28.7 0 0 28.7 0 64v288c0 35.3 28.7 64 64 64h96v84c0 7.1 5.8 12 12 12 2.4 0 4.9-.7 7.1-2.4L304 416h144c35.3 0 64-28.7 64-64V64c0-35.3-28.7-64-64-64zm32 352c0 17.6-14.4 32-32 32H293.3l-8.5 6.4L192 460v-76H64c-17.6 0-32-14.4-32-32V64c0-17.6 14.4-32 32-32h384c17.6 0 32 14.4 32 32v288zM280 240H136c-4.4 0-8 3.6-8 8v16c0 4.4 3.6 8 8 8h144c4.4 0 8-3.6 8-8v-16c0-4.4-3.6-8-8-8zm96-96H136c-4.4 0-8 3.6-8 8v16c0 4.4 3.6 8 8 8h240c4.4 0 8-3.6 8-8v-16c0-4.4-3.6-8-8-8z"
        fill="currentColor"
        style={Object {}}
      />
    </svg>
  </div>
  <div
    className="c1"
    size="1rem"
  >
    <svg
      aria-hidden="true"
      className="svg-inline--fa fa-browser fa-w-16 fa-1x "
      data-icon="browser"
      data-prefix="fal"
      focusable="false"
      role="img"
      style={Object {}}
      viewBox="0 0 512 512"
      xmlns="http://www.w3.org/2000/svg"
    >
      <path
        d="M464 32H48C21.5 32 0 53.5 0 80v352c0 26.5 21.5 48 48 48h416c26.5 0 48-21.5 48-48V80c0-26.5-21.5-48-48-48zM32 80c0-8.8 7.2-16 16-16h48v64H32V80zm448 352c0 8.8-7.2 16-16 16H48c-8.8 0-16-7.2-16-16V160h448v272zm0-304H128V64h336c8.8 0 16 7.2 16 16v48z"
        fill="currentColor"
        style={Object {}}
      />
    </svg>
  </div>
</div>
`;

exports[`Icon renders correctly 1`] = `
.c0 {
  display: -webkit-box;
  display: -webkit-flex;
  display: -ms-flexbox;
  display: flex;
  -webkit-flex-direction: row;
  -ms-flex-direction: row;
  flex-direction: row;
  -webkit-box-pack: center;
  -webkit-justify-content: center;
  -ms-flex-pack: center;
  justify-content: center;
  -webkit-align-items: center;
  -webkit-box-align: center;
  -ms-flex-align: center;
  align-items: center;
  width: 1rem;
  height: 1rem;
  font-size: 1rem;
}

<div
  className="c0"
>
  <svg
    aria-hidden="true"
    className="svg-inline--fa fa-users fa-w-20 fa-1x "
    data-icon="users"
    data-prefix="fal"
    focusable="false"
    role="img"
    style={Object {}}
    viewBox="0 0 640 512"
    xmlns="http://www.w3.org/2000/svg"
  >
    <path
      d="M544 224c44.2 0 80-35.8 80-80s-35.8-80-80-80-80 35.8-80 80 35.8 80 80 80zm0-128c26.5 0 48 21.5 48 48s-21.5 48-48 48-48-21.5-48-48 21.5-48 48-48zM320 256c61.9 0 112-50.1 112-112S381.9 32 320 32 208 82.1 208 144s50.1 112 112 112zm0-192c44.1 0 80 35.9 80 80s-35.9 80-80 80-80-35.9-80-80 35.9-80 80-80zm244 192h-40c-15.2 0-29.3 4.8-41.1 12.9 9.4 6.4 17.9 13.9 25.4 22.4 4.9-2.1 10.2-3.3 15.7-3.3h40c24.2 0 44 21.5 44 48 0 8.8 7.2 16 16 16s16-7.2 16-16c0-44.1-34.1-80-76-80zM96 224c44.2 0 80-35.8 80-80s-35.8-80-80-80-80 35.8-80 80 35.8 80 80 80zm0-128c26.5 0 48 21.5 48 48s-21.5 48-48 48-48-21.5-48-48 21.5-48 48-48zm304.1 180c-33.4 0-41.7 12-80.1 12-38.4 0-46.7-12-80.1-12-36.3 0-71.6 16.2-92.3 46.9-12.4 18.4-19.6 40.5-19.6 64.3V432c0 26.5 21.5 48 48 48h288c26.5 0 48-21.5 48-48v-44.8c0-23.8-7.2-45.9-19.6-64.3-20.7-30.7-56-46.9-92.3-46.9zM480 432c0 8.8-7.2 16-16 16H176c-8.8 0-16-7.2-16-16v-44.8c0-16.6 4.9-32.7 14.1-46.4 13.8-20.5 38.4-32.8 65.7-32.8 27.4 0 37.2 12 80.2 12s52.8-12 80.1-12c27.3 0 51.9 12.3 65.7 32.8 9.2 13.7 14.1 29.8 14.1 46.4V432zM157.1 268.9c-11.9-8.1-26-12.9-41.1-12.9H76c-41.9 0-76 35.9-76 80 0 8.8 7.2 16 16 16s16-7.2 16-16c0-26.5 19.8-48 44-48h40c5.5 0 10.8 1.2 15.7 3.3 7.5-8.5 16.1-16 25.4-22.4z"
      fill="currentColor"
      style={Object {}}
    />
  </svg>
</div>
`;
